### ISSUE TITLE
Plugins: npm plugin

### ISSUE THREAD

**Issue by dandelionmane** [sourcecred/sourcecred#42]
Let's create a plugin with `package.json` files as nodes, and `npm dependencies` as edges.

I believe this will give a very interesting view into meta-project dynamics, and developing it early will force us to keep the meta-scale in mind as we develop tooling and visualizations. Also, it should be fairly easy to discover this data. We can start by exploring all of SourceCred's transitive dependencies.

**Comment by Beanow** [comment]
Would be very interesting!

> Also, it should be fairly easy to discover this data.

Aaah, allow me to share a little doom scenario for the meta-scale then.

The other day I found out about sourcecred and js-ipfs, which is in the prototype, happens to be something I contributed a little bit to. Comments and stuff are all there, though probably the most relevant thing I did for js-ipfs at the time was improve performance of an upstream dependency `pull-block`. As noted in the release issue here https://github.com/ipfs/js-ipfs/issues/1095
Not an easy contribution to find though! It *was* 2 steps to even get to this dependency (`js-ipfs` > `js-ipfs-unixfs-engine` (split in different packages now) > `pull-block`). But you won't find it in the latest version as it was replaced with another library for performance. https://github.com/ipfs/js-ipfs-unixfs-importer/pull/12
Also, the lock files aren't committed. So to discover cases like this, you'd need to analyze every version of the package.json file. And potentially every version of the dependencies' dependencies too, and so on. Pretty soon you'll need a local npm cache.

I'm not proposing this should be included, just a concrete example you might use to see how far down the rabbit hole this should go.

**Comment by decentralion** [comment]
Yeah, you're right @Beanow that this is a very deep rabbit hole once we want to attribute cred at each moment in the history. I was thinking that we could have a standalone "JS cred instance" which just runs PageRank on the npm dependency graph at the present moment in time. But that wouldn't integrate cleanly with any individual project's cred, because it's missing the history. The example you gave is a good illustration of that issue.

**Comment by Beanow** [comment]
Well not to be discouraged. I presume the design goal was not to have a perfect quantifying algorithm. So I hope it's useful input for defining new milestones and having ever closer approximations :smile:
